### PR TITLE
add thandle create from content flex and thandle move

### PR DESCRIPTION
**Pull request by anporumb**

### CODE REVIEW

File: devdoc/thandle_requirements.md [hunk 1]
```diff
@@ -165,22 +165,54 @@ Given a previously constructed `THANDLE(T)`, `THANDLE_GET_T(T)` reeturns a point
...
+###  THANDLE_CREATE_FROM_CONTENT_MACRO(T)
 ```c
 THANDLE_CREATE_FROM_CONTENT_MACRO(T)
 THANDLE(T) THANDLE_CREATE_FROM_CONTENT(T)(const T* source, void(*dispose)(T*), int(*copy)(T* destination, const T* source))
 ```
 
-Given a previously existing T, `THANDLE_CREATE_FROM_CONTENT` will copy `T`'s content and return a `THANDLE(T)`.
+`THANDLE_CREATE_FROM_CONTENT_MACRO` is the same as `THANDLE_CREATE_FROM_CONTENT_FLEX_MACRO` where `get_sizeof` function returns what the C's sizeof operator evaluates to.
+
+**SRS_THANDLE_02_032: [** `THANDLE_CREATE_FROM_CONTENT` returns what `THANDLE_CREATE_FROM_CONTENT_FLEX(T)(source, dispose, copy, THANDLE_GET_SIZEOF(T));` returns. **]**
```
Comment (mattdurak) [inline devdoc/thandle_requirements.md:198]: >THANDLE_GET_SIZEOF [](start = 129, length = 18)

I don't see this anywhere else in the doc (but I can assume it's really `sizeof` based on the description above).

File: inc/azure_c_util/thandle.h [hunk 2]
```diff
@@ -54,9 +54,15 @@
...
 #define THANDLE_MALLOC_WITH_EXTRA_SIZE(T) MU_C2(T,_MALLOC_WITH_EXTRA_SIZE)
 
-/*given a previous type T, THANDLE_CREATE_FROM_CONTENT introduces a new name for the function that makes of copy of T and returns a THANDLE to the copy*/
+/*given a previous type T, THANDLE_CREATE_FROM_CONTENT introduces a new name for the function that makes of copy of T and returns a THANDLE to the caller*/
 #define THANDLE_CREATE_FROM_CONTENT(T) MU_C2(T,_CREATE_FROM_CONTENT)
 
+/*given a previous type T (usually a struct with a flexible array member), THANDLE_CREATE_FROM_CONTENT_FLEX introduces a new name for the function that makes of copy of T and returns a THANDLE to the caller*/
+#define THANDLE_CREATE_FROM_CONTENT_FLEX(T) MU_C2(T,_CREATE_FROM_CONTENT_FLEX)
+
+/*given a previous type T (usually a struct with a flexible array member), THANDLE_GET_SIZEOF introduces a new name for a user functiont that returns the sizeof of T */
```
Comment (mattdurak) [inline inc/azure_c_util/thandle.h:63]: >functiont [](start = 127, length = 9)

functiont

**Comment by mattdurak** [comment]
> `THANDLE_TYPE_DECLARE` introduces several functions that can be used with `THANDLE(T)` type. These are `THANDLE_DEC_REF(T)`, `THANDLE_INC_REF(T)`, `THANDLE_ASSIGN(T)`, `THANDLE_INITIALIZE(T)`.

THANDLE_MOVE

---
Refers to: devdoc/thandle_requirements.md:52 in 54214ad. [](commit_id = 54214ad598ab39a12b98cc38620050984fbda139, deletion_comment = False)

File: inc/azure_c_util/thandle.h (hunk 2 above)
Comment (anporumb) [reply inc/azure_c_util/thandle.h:63]: changed to function

---
In reply to: [411489221](https://github.com/Azure/azure-c-util/pull/28#discussion_r411489221) [](ancestors = 411489221)

**Comment by anporumb** [comment]
> `THANDLE_TYPE_DECLARE` introduces several functions that can be used with `THANDLE(T)` type. These are `THANDLE_DEC_REF(T)`, `THANDLE_INC_REF(T)`, `THANDLE_ASSIGN(T)`, `THANDLE_INITIALIZE(T)`.

added

---
In reply to: [616642361](https://github.com/Azure/azure-c-util/pull/28#issuecomment-616642361) [](ancestors = 616642361)

---
Refers to: devdoc/thandle_requirements.md:52 in 54214ad. [](commit_id = 54214ad598ab39a12b98cc38620050984fbda139, deletion_comment = False)

File: devdoc/thandle_requirements.md (hunk 1 above)
Comment (anporumb) [reply devdoc/thandle_requirements.md:206]: added an explicit chapter for THANDLE_GET_SIZEOF

---
In reply to: [411487113](https://github.com/Azure/azure-c-util/pull/28#discussion_r411487113) [](ancestors = 411487113)

**Pull request merged**